packages/video_player/video_player_windows/windows/MediaEngineWrapper.cpp

```diff
@@ -0,0 +1,506 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <windows.h>
+
+// Include prior to C++/WinRT Headers
+#include <wil/cppwinrt.h>
+
+// Windows Implementation Library
+#include <wil/resource.h>
+
+// MediaFoundation headers
+#include <Audioclient.h>
+#include <d3d11.h>
+#include <dxgi1_2.h>
+#include <mfapi.h>
+#include <mferror.h>
+#include <mfmediaengine.h>
+
+// STL headers
+#include <functional>
+#include <memory>
+
+#include "MediaEngineWrapper.h"
+
+namespace video_player_windows {
+
+namespace {
+
+// Implementation of IMFMediaEngineNotify's interface, to handle MediaEngine events.
+class MediaEngineCallbackHelper
+    : public winrt::implements<MediaEngineCallbackHelper,
+                               IMFMediaEngineNotify> {
+ public:
+  MediaEngineCallbackHelper(
+      std::function<void()> on_loaded_cb, MediaEngineWrapper::ErrorCB error_cb,
+      MediaEngineWrapper::BufferingStateChangeCB buffering_state_change_cb,
+      std::function<void()> playback_ended_cb,
+      std::function<void()> time_update_cb)
+      : on_loaded_cb_(on_loaded_cb),
+        error_cb_(error_cb),
+        buffering_state_change_cb_(buffering_state_change_cb),
+        playback_ended_cb_(playback_ended_cb),
+        time_update_cb_(time_update_cb) {
+    // Ensure that callbacks are valid
+    THROW_HR_IF(E_INVALIDARG, !on_loaded_cb_);
+    THROW_HR_IF(E_INVALIDARG, !error_cb_);
+    THROW_HR_IF(E_INVALIDARG, !buffering_state_change_cb_);
+    THROW_HR_IF(E_INVALIDARG, !playback_ended_cb_);
+    THROW_HR_IF(E_INVALIDARG, !time_update_cb_);
+  }
+  virtual ~MediaEngineCallbackHelper() = default;
+
+  void DetachParent() {
```

Comment:
Who calls this? Should the destructor call thsi?

This doesn't appear to be on the `IMFMediaEngineNotify` interface and I couldn't find any references in this code base. Consider adding a comment here.

Reply:
Nice catch, this was missing. Should be called from the dctor.

Reply:
Fixed.